### ISSUE TITLE
Release v14.0.0

### ISSUE THREAD

**Issue by leebyron** [graphql/graphql-js#1383]
Let's coordinate on releasing this final version, @IvanGoncharov and @mjmahone.

I believe #1382 is the final blocker based on previous discussions I've had with you all about what we'd like to include. If there's anything else we should have blocking this release let me know.

Next steps are to write up the full release notes with special mentions of breaking changes, then finally push the version bump and commit tag.

**Comment by IvanGoncharov** [comment]
@leebyron There is a couple issue I think we need to dissucss:
1. After #1373 merged we still have a few related issues:
    1. `validateSchema` doesn't use nodes from new `extensionASTNodes` properties.
    2. `extend scalar` is not supported
1. I also think it's very critical to add validation of directives in SDL since as [noted here](https://github.com/facebook/graphql/issues/429#issuecomment-392990065) no one validates them correctly at the moment. That will lead to a situation very similar to the recent `Allow interfaces to have no implementors`. Since spec defines set of validation rules but no one implements them. Probably we shouldn't enforce `Directives Are Unique Per Location` until https://github.com/facebook/graphql/issues/429 resolved but all other validation should be implemented e.g. require schema to contain applied directives, correct directive arguments, etc.
1. Should we switch from IE9 to IE11 especially since IE9 & IE10 not [fully supported by babel](https://babeljs.io/docs/usage/caveats/#internet-explorer).
Don't know a lot about ios but maybe it also worth to update. 
https://github.com/graphql/graphql-js/blob/f2070c8169617b949b17c1929d27e46454a1d90a/.babelrc.js#L7-L8
1. I think we should use this oportunity to clarify on deprecated functionality. I propose the following:
    1. Drop support for [onOperation, onFragment, onField](https://github.com/graphql/graphql-js/blob/bd8aa362efc2dcafb54ed1b7286577a6ee7b4c59/src/type/introspection.js#L95-L117) since they were deprecated two years ago.
    2. Explicitly say that following functions would be removed in `15.0.0` release [isValidJSValue](https://github.com/graphql/graphql-js/blob/2536f144e30fb33aadec3916dfb728043fae573b/src/utilities/isValidJSValue.js#L14),[isValidLiteralValue](https://github.com/graphql/graphql-js/blob/17a0bfd5292f39cafe4eec5b3bd0e22514243b68/src/utilities/isValidLiteralValue.js#L23),[introspectionQuery](https://github.com/graphql/graphql-js/blob/b333b30b238d040d6a8cf959d54f037ee26ae2c6/src/utilities/index.js#L14).
    3. Explicitly say that support for `legacy SDL syntax` would be removed in `16.0.0`. Also we can recommend all 3rd-party tools and libraries to stop supporting it by default and require user to explicitly specify that he wants use `legacy SDL syntax` through configuration/option.

**Comment by IvanGoncharov** [comment]
@leebyron I also forget to mention one more breaking change which is not super important but a huge quality of life improvement. `parseLiteral` should throw on invalid values similar to `parseValue`.
https://github.com/graphql/graphql-js/blob/f2070c8169617b949b17c1929d27e46454a1d90a/src/type/scalars.js#L56-L64
Same goes for enums:
https://github.com/graphql/graphql-js/blob/f2070c8169617b949b17c1929d27e46454a1d90a/src/type/definition.js#L1105-L1113

**Comment by leebyron** [comment]
@IvanGoncharov want to open a new issue for throwing from parseLiteral? Not sure I totally understand the use case and want to discuss further but don't want to hijack this issue

**Comment by leebyron** [comment]
To follow up on point 3, changing "ie 9" to "ie 11" has no impact on compiled output (tested with `diff -r`), only removing both `ie` and `ios` entries causes any change - for example then `const` is not compiled to `var` - however I don't think this is something we want to do, so I'll suggest leaving that configuration as is for now.

**Comment by leebyron** [comment]
I created new tasks for the remaining issues you brought up and created a milestone so we keep track of them: https://github.com/graphql/graphql-js/milestone/1

If you (or anyone else who might be reading this? @mjmahone?) is interested in helping work through these that would be a huge help in getting this release out this week.

**Comment by alexmylonas** [comment]
Excuse my ignorance, but why you jumped from v0.13 -> v14? Is this misname or is this a common practice, just curious!

**Comment by jgcmarins** [comment]
@alexmylonas -> https://github.com/graphql/graphql-js/issues/1005

**Comment by alexmylonas** [comment]
Thanks for the response. Solid read and discussion

**Comment by PowerKiKi** [comment]
This issue should probably be part of milestone v14.0.0.

**Comment by nakleiderer** [comment]
I'm new and likely missing some context, but wouldn't the next version be v0.14.0?

**Comment by mjmahone** [comment]
I'm onboard with cutting the 14.0.0 release. I don't think there's anything we're waiting for that can't be pushed to a 14.1 release, but I want to make sure we have consensus with @IvanGoncharov too.

**Comment by IvanGoncharov** [comment]
@leebyron As for #1389 we currently validating that:
* directive is known
* it used on correct location
* all directive args are known
* all required arguments specified
* no directive duplication on the same location (but without checking `extend`s)
So we not checking only:
https://github.com/graphql/graphql-js/blob/master/src/validation/rules/ValuesOfCorrectType.js

My idea was to force developers to specify they use inside SDL so I think current state satisfy this. We can adapt `ValuesOfCorrectType` to SDL in `14.1` release since it's not breaking change just an additional validation.

There is two small issue I want to fix before `14.0.0` release:
* #1494
* And drop support for `undefined` as enum value since it taints not only enum values but also `defaultValue` and all code that work with values. 90% of this library doesn't handle `undefined` as valid value. For example any function that do schema transformation will unintentionally define default value for every arg simply by writing code like this:
https://github.com/graphql/graphql-js/blob/f4dee2803c22f7e5fde913b9743cc6de345bb871/src/utilities/lexicographicSortSchema.js#L86-L91
\+ by adding bunch of `hasOwnProperty` all over the place we introduce a lot of JS-specific code into implementation and bunch of test cases for almost every function.

I don't think anyone really using this feature but it creates a lot of inconcistency in this library.
I will prepare PR tomorrow morning and it the last thing I want to disscuss before `14.0.0`

@leebyron If you have time it would be great if you could review PRs with breaking changes: https://github.com/graphql/graphql-js/pulls?q=is%3Apr+sort%3Aupdated-desc+is%3Aclosed+label%3A%22semver+major%22
To be sure that we didn't miss anything major before we freeze public API for `v14.x.x`.

**Comment by mjmahone** [comment]
> And drop support for undefined as enum value

Given, as you've pointed out, that 90% of our code can't handle it as-is, I don't think it would be a real breaking change either way to do one of: 
- drop support entirely (because no one could use it correctly anyways)
- do the hard work to actually support it

I'm really happy with the state things are in right now, but I will try to do the release tomorrow morning around 13:00 UTC to give time for all of us to comb over things one more time.

**Comment by IvanGoncharov** [comment]
@mjmahone Sound good 👍 
Since it's UTC+3 for me I will have time to go over all open issues one more time to see that we didn't miss anything.
Do you need any help with release notes?
I can create a list of commits and group them into by bug/feature.
We also need to write short instruction about two ways how you can specifying directives for your SDL.

**Comment by mjmahone** [comment]
Alright, I'll do the release in ~10 minutes. @IvanGoncharov I know you've put up some last minute PRs: anything worth waiting for?

**Comment by IvanGoncharov** [comment]
> anything worth waiting for?

@mjmahone No, I think we are good to go 🏁🏎
I think we could disscuss `undefined` as enum values beetween RC and `14.0.0` release.

Do you need any help with release notes?

**Comment by mjmahone** [comment]
@IvanGoncharov plan is to do a full 14.0.0 release, not an RC. This release has been baking for long enough, let's just rip off the band-aid.

I think I should be OK aggregating my previous RC release notes and the changes since then.

**Comment by IvanGoncharov** [comment]
@mjmahone Changes in directive validation can potentially break Relay or Facebook tools. Based on previous experience I'm worried that it will provoke reverts.
Can you please ensure that it doesn't break anything at Facebook.
Maybe it will make sense to create intermediate RC for a couple of days and agree that we don't merge anything in `master`.

**Comment by mjmahone** [comment]
I'm responsible for merging it into Facebook's tools, so I'll make sure. Using an intermediate RC doesn't actually help though: we can't do the Relay upgrade to an RC, though I can test a bit with an RC and find some potential issues (can't merge it into trunk as long as it's an RC, though).

If we need to do a quick, follow-on patch release, we will, but I don't think Relay currently has much potential for directive validation issues, and if it does I have the ability to fix Relay forward.

**Comment by IvanGoncharov** [comment]
@mjmahone In that case, I'm OK with the release.

**Comment by mjmahone** [comment]
Release: https://github.com/graphql/graphql-js/releases/tag/v14.0.0

**Comment by IvanGoncharov** [comment]
@leebyron Thanks Lee 

@mjmahone It also missing #1284 and migration strategy:
https://github.com/graphql/graphql-js/pull/1284#issuecomment-377811537

**Comment by mjmahone** [comment]
That was one of the easiest major releases to upgrade in my living memory: there were flow errors anywhere there were problems, and it was immediately clear what the issue actually was. Some of the ease may be due to having fixed a lot of super-clowniness earlier when I was testing the "rc" release.

As a summary of things we ran into internally at FB with the upgrade:
- A few queries were of the form `query Foo($requiredVariable: ID = null)`, where the usage of `$requiredVariable` is as a non-nullable argument. This is now a validation error! 🙌
- Had to change flow types that were `TypeSystemDefinitionNode` to `TypeSystemDefinitionNode | TypeSystemExtensionNode`, because we split extensions from definitions. These functions probably deserve an audit, and some of them should be split.
- Relay had to do strict-object-typing of a `Location` type.
- A few places we check for `definition.name`, we had to switch `if (node.kind === SCHEMA_DEFINITION)` to `if (node.kind === SCHEMA_DEFINITION || node.kind === SCHEMA_EXTENSION)` to cover the two type system definitions/extensions that have no `name` field.
- A few places I had to explicitly give a `const someDefinition: OperationDefinitionNode = ...` instead of having it be implicit.

**Comment by leebyron** [comment]
That’s great to hear! For your last point, I’d think AST nodes would be the easiest to type implicitly since they have the “kind” field, making them a disjoint union. No big deal to add an explicit type, but that seems like a warning light to me that something else is wrong

**Comment by IvanGoncharov** [comment]
@mjmahone I agree with Lee it either problem with our typings or Flow.
Is it part of open source project? If not can you please split it out into separate example so we can together figure out what's the problem?

**Comment by mjmahone** [comment]
I'm pretty sure it's sketchy code we're doing internally that no one should copy. It was really easy to fix.

It's something like:
```
function thisIsAProblem(): DocumentNode {
  const someOperationAST: OperationDefinitionNode = {
    kind: OPERATION_DEFINITION,
    ...
  };
  return modifiesDocument({
    kind: DOCUMENT,
    definitions: [{
      ...someOperationAST,
      variableDefinitions: createsVariables(),
    }],
  }));
}

function modifiesDocument(doc: DocumentNode): DocumentNode { ... }
function createsVariables(): Array<VariableDefinitionNode> { ... }
```

Flow can't figure out how to disambiguate this. The error is:
```
Could not decide which case to select. Since case 1 [1] may work but if it doesn't case 3 [2] looks promising too. To
fix add a type annotation to .directives [3], to .kind [3], to .loc [3], to .name [3], to .operation [3], to
.selectionSet [3], or to .variableDefinitions [3].
```

However, it's trivially solved by doing:
```
function thisIsNotAProblem(): DocumentNode {
  const someOperationAST: OperationDefinitionNode = {
    kind: OPERATION_DEFINITION,
    ...
  };
  const modifiedOperation: OperationDefinitionNode = {
    ...someOperationAST,
    variableDefinitions: createsVariables(),
  };
  return modifiesDocument({
    kind: DOCUMENT,
    definitions: [modifiedOperation],
  }));
```

**Comment by IvanGoncharov** [comment]
@mjmahone Thanks for sharing
Agree, it's probably too much for a flow since it's hard to figure out exact type after spread and also `definitions` accept the long list of possible types.

**Comment by IvanGoncharov** [comment]
Also `OperationDefinitionNode` is not exact type and it probably makes it harder to figure out the type of spread:
https://github.com/graphql/graphql-js/blob/6eec989f89a81bf81b8f0cdb6ceebf55d6edd93a/src/language/ast.js#L210-L218
